KeyError: 'destination' in fill()

Hello !

Getting a new KeyError with 2.5.0:

```
 djblockchain/tezos.py:140: in get_client
    operation = client.reveal().autofill().sign().inject()
/usr/lib/python3.8/site-packages/pytezos/tools/docstring.py:67: in __call__
    return method(self.class_instance, *args, **kwargs)
/usr/lib/python3.8/site-packages/pytezos/operation/group.py:167: in autofill
    opg = self.fill()
/usr/lib/python3.8/site-packages/pytezos/tools/docstring.py:67: in __call__
    return method(self.class_instance, *args, **kwargs)
/usr/lib/python3.8/site-packages/pytezos/operation/group.py:122: in fill
    contents=list(map(fill_content, self.contents)),
/usr/lib/python3.8/site-packages/pytezos/operation/group.py:118: in fill_content
    content[k] = v(content) if callable(v) else v
/usr/lib/python3.8/site-packages/pytezos/operation/group.py:109: in <lambda>
    'fee': lambda x: str(default_fee(x)),
/usr/lib/python3.8/site-packages/pytezos/operation/fees.py:32: in default_fee
    consumed_gas=default_gas_limit(content),
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
content = {'counter': '1', 'fee': '0', 'gas_limit': '0', 'kind': 'reveal', ...}
    def default_gas_limit(content) -> int:
        """ Get default gas limit by operation kind.
    
        :param content: operation content {..., "kind": "transaction", ... }
        """
        values = {
            'reveal': 10000,
            'delegation': 10000,
            'origination': hard_gas_limit_per_operation if content.get('script') else 10000,
>           'transaction': hard_gas_limit_per_operation if content['destination'].startswith('KT') else 10207
        }
E       KeyError: 'destination'
/usr/lib/python3.8/site-packages/pytezos/operation/fees.py:46: KeyError
```

No problem with <2.5.0

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -1,6 +1,6 @@
 [tool.poetry]
 name = "pytezos"
-version = "2.5.0"
+version = "2.5.1"
 description = "Python toolkit for Tezos"
 license = "MIT"
 authors = ["Michael Zaikin <mz@baking-bad.org>", "Arthur Breitman", "Roman Serikov"]

--- a/pytezos/__init__.py
+++ b/pytezos/__init__.py
@@ -14,6 +14,7 @@ from pytezos.rpc import RpcProvider, localhost, mainnet, babylonnet, zeronet
 from pytezos.rpc.errors import *
 from pytezos.crypto import Key
 from pytezos.proto import Proto
+from pytezos.michelson.micheline import Unit
 from pytezos.michelson.contract import Contract
 from pytezos.michelson.formatter import format_timestamp
 from pytezos.client import PyTezosClient

--- a/pytezos/michelson/docstring.py
+++ b/pytezos/michelson/docstring.py
@@ -5,7 +5,7 @@ from pytezos.michelson.micheline import Schema, is_optional
 core_types = ['string', 'int', 'bool']
 domain_types = {
     'nat': 'int  /* Natural number */',
-    'unit': 'None /* Void */',
+    'unit': 'Unit || None /* Void */',
     'bytes': 'string  /* Hex string */ ||\n\tbytes  /* Python byte string */',
     'timestamp': 'int  /* Unix time in seconds */ ||\n\tstring  /* Formatted datetime `%Y-%m-%dT%H:%M:%SZ` */',
     'mutez': 'int  /* Amount in `utz` (10^-6) */ ||\n\tDecimal  /* Amount in `tz` */',

--- a/pytezos/michelson/grammar.py
+++ b/pytezos/michelson/grammar.py
@@ -195,4 +195,6 @@ class MichelsonParser(object):
         :param code: Michelson source
         :returns: Micheline expression
         """
+        if len(code) > 0 and code[0] == '(' and code[-1] == ')':
+            code = code[1:-1]
         return self.parser.parse(code)

--- a/pytezos/michelson/micheline.py
+++ b/pytezos/michelson/micheline.py
@@ -12,6 +12,7 @@ from pytezos.repl.parser import parse_expression, dispatch_core_map
 
 Nested = namedtuple('Nested', ['prim', 'args'])
 Schema = namedtuple('Schema', ['metadata', 'bin_types', 'bin_names', 'json_to_bin'])
+Unit = None
 
 
 class Pair(tuple):
@@ -230,7 +231,7 @@ def parse_micheline(val_expr, type_expr, schema: Schema, bin_root='0'):
             res = {schema.bin_names[arg_path]: val[0]} if is_leaf else val[0]
             return next(iter(res)) if bin_type == 'enum' else res
         elif bin_type == 'unit':
-            return None
+            return Unit
         elif bin_type == 'lambda':
             return micheline_to_michelson(val)
         elif bin_type == 'timestamp':
@@ -419,8 +420,6 @@ def michelson_to_micheline(data, parser=None):
     """
     if parser is None:
         parser = michelson_parser()
-    if data[0] == '(' and data[-1] == ')':
-        data = data[1:-1]
     return parser.parse(data)
 
 

--- a/pytezos/operation/fees.py
+++ b/pytezos/operation/fees.py
@@ -43,7 +43,7 @@ def default_gas_limit(content) -> int:
         'reveal': 10000,
         'delegation': 10000,
         'origination': hard_gas_limit_per_operation if content.get('script') else 10000,
-        'transaction': hard_gas_limit_per_operation if content['destination'].startswith('KT') else 10207
+        'transaction': hard_gas_limit_per_operation if content.get('destination', '').startswith('KT') else 10207
     }
     return values.get(content['kind'])
 
@@ -57,6 +57,6 @@ def default_storage_limit(content):
         'reveal': 0,
         'delegation': 0,
         'origination': hard_storage_limit_per_operation if content.get('script') else 10207,
-        'transaction': hard_storage_limit_per_operation if content['destination'].startswith('KT') else 257
+        'transaction': hard_storage_limit_per_operation if content.get('destination', '').startswith('KT') else 257
     }
     return values.get(content['kind'])

--- a/pytezos/repl/helpers.py
+++ b/pytezos/repl/helpers.py
@@ -2,10 +2,11 @@ from os.path import isfile
 
 from pytezos import Contract
 from pytezos.interop import Interop
-from pytezos.encoding import is_kt
+from pytezos.encoding import is_kt, forge_timestamp
 from pytezos.repl.control import instruction, do_interpret
 from pytezos.repl.context import Context
-from pytezos.repl.parser import get_int, get_string, get_bool, parse_prim_expr, restore_entry_expr, assert_expr_equal
+from pytezos.repl.parser import get_int, get_string, get_bool, parse_prim_expr, restore_entry_expr, assert_expr_equal, \
+    dispatch_core_map
 from pytezos.repl.types import Pair, Mutez, Address, ChainID, Timestamp, assert_stack_type, List, Operation
 
 networks = ['mainnet', 'zeronet', 'babylonnet', 'carthagenet']
@@ -123,7 +124,7 @@ def do_patch(ctx: Context, prim, args, annots):
     if key in ['AMOUNT', 'BALANCE']:
         res = Mutez(get_int(args[1]))
     elif key == 'NOW':
-        res = Timestamp(get_int(args[1]))
+        res = Timestamp(dispatch_core_map(args[1], {'int': int, 'string': forge_timestamp}))
     elif key in ['SOURCE', 'SENDER']:
         res = Address.new(get_string(args[1]))
     elif key == 'CHAIN_ID':

--- a/pytezos/repl/parser.py
+++ b/pytezos/repl/parser.py
@@ -87,7 +87,10 @@ def dispatch_core_map(val_expr, mapping: dict):
     assert act_type in mapping, f'expected {expected}, got {act_type}'
     res = mapping[act_type]
     if callable(res):
-        return res(val)
+        try:
+            return res(val)
+        except ValueError as e:
+            raise MichelsonTypeCheckError.init(str(e), act_type)
     else:
         return res
 

--- a/tests/test_decode.py
+++ b/tests/test_decode.py
@@ -3,6 +3,7 @@ from parameterized import parameterized
 
 from pytezos.michelson.converter import michelson_to_micheline
 from pytezos.michelson.contract import ContractStorage, ContractParameter
+from pytezos.michelson.grammar import MichelsonParser
 
 storage_data = [(
     """Pair (Pair "tz1YPSCGWXwBdTncK2aCctSZAXWvGsGwVJqU" True) 
@@ -96,3 +97,7 @@ class TestComparable(TestCase):
         decoded = parameter.decode(source)
         encoded = parameter.encode(decoded, entrypoint='default')
         self.assertEqual(source, encoded['value'])
+
+    def test_wrapped_expr(self):
+        p = MichelsonParser()
+        p.parse('(Pair 1 2)')

--- a/tests/test_repl_regression.py
+++ b/tests/test_repl_regression.py
@@ -1,6 +1,6 @@
 from unittest import TestCase
 
-from pytezos.repl.interpreter import Interpreter, StackItem
+from pytezos.repl.interpreter import Interpreter, StackItem, MichelsonRuntimeError
 from pytezos.repl.control import assert_stack_type
 from pytezos.repl.types import Option
 
@@ -21,3 +21,13 @@ class TestREPLRegression(TestCase):
         top = i.ctx.stack[0]  # type: Option
         assert_stack_type(top, Option)
         self.assertTrue(top.is_none())
+
+    def test_now_failed_parse_exception_type(self):
+        i = Interpreter()
+        with self.assertRaises(MichelsonRuntimeError):
+            i.execute('PUSH timestamp "2020-06-17 10:00:00"')
+
+    def test_patch_now_str(self):
+        i = Interpreter()
+        i.execute('PATCH NOW "2020-06-17T10:00:00Z"')
+        i.execute('PATCH NOW 1593082373')

